core/src/main/java/org/apache/calcite/rel/externalize/RelDotWriter.java

```diff
@@ -0,0 +1,210 @@
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Utility to dump a rel node plan in dot format.
+ */
+public class RelDotWriter extends RelWriterImpl {
+
+  //~ Instance fields --------------------------------------------------------
+
+  /**
+   * Adjacent list of the plan graph.
+   */
+  private final Map<RelNode, List<RelNode>> outArcTable = new LinkedHashMap<>();
+
+  private Map<RelNode, String> nodeLabels = new HashMap<>();
+
+  private Multimap<RelNode, String> nodeStyles = HashMultimap.create();
+
+  private final WriteOption option;
+
+  //~ Constructors -----------------------------------------------------------
+
+  public RelDotWriter(
+      PrintWriter pw, SqlExplainLevel detailLevel,
+      boolean withIdPrefix) {
+    this(pw, detailLevel, withIdPrefix, new WriteOption());
+  }
+
+  public RelDotWriter(
+      PrintWriter pw, SqlExplainLevel detailLevel,
+      boolean withIdPrefix, WriteOption option) {
+    super(pw, detailLevel, withIdPrefix);
+    this.option = option;
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  protected void explain_(RelNode rel,
+                          List<Pair<String, Object>> values) {
+    // get inputs
+    List<RelNode> inputs = getInputs(rel);
+    outArcTable.put(rel, inputs);
+
+    // generate node label
+    StringBuilder relDesc = new StringBuilder();
+    populateRelNodeLabel(rel, values, relDesc);
+    nodeLabels.put(rel, option.formatNodeLabel(relDesc.toString()));
+
+    if (option.highlightNode(rel)) {
+      nodeStyles.put(rel, "bold");
+    }
+
+    explainInputs(inputs);
+  }
+
+  private List<RelNode> getInputs(RelNode parent) {
+    return parent.getInputs().stream().map(child -> {
+      if (child instanceof HepRelVertex) {
+        return ((HepRelVertex) child).getCurrentRel();
+      } else if (child instanceof RelSubset) {
+        RelSubset subset = (RelSubset) child;
+        return Util.first(subset.getBest(), subset.getOriginal());
+      } else {
+        return child;
+      }
+    }).collect(Collectors.toList());
+  }
+
+  private void explainInputs(List<RelNode> inputs) {
+    for (RelNode input : inputs) {
+      if (input == null || nodeLabels.containsKey(input)) {
+        continue;
+      }
+      input.explain(this);
+    }
+  }
+
+  public RelWriter done(RelNode node) {
+    int numOfVisitedNodes = nodeLabels.size();
+    super.done(node);
+    if (numOfVisitedNodes == 0) {
+      // When we enter this method call, no node
+      // has been visited. So the current node must be the root of the plan.
+      // Now we are exiting the method, all nodes in the plan
+      // have been visited, so it is time to dump the plan
+
```

Comment:
Minor: it is time to dump the plan -> `it is time to dump the plan.`

Reply:
Nice catch. Thank you.